rest_server/database/interface.py

```diff
@@ -0,0 +1,211 @@
+                404,
+                log_message=f"Document Not Found: {coll} document ({scan_id})",
+            )
+        scandc = from_dict(scandc_type, doc)
+        LOGGER.debug(f"found: ({coll=}) doc {scandc}")
+        return scandc  # type: ignore[no-any-return]  # mypy internal bug
+
+    async def _upsert(self, coll: str, scandc: S) -> S:
+        """Insert/update the doc."""
+        LOGGER.debug(f"replacing: ({coll=}) doc with {scandc=}")
+        doc = await self._collections[coll].find_one_and_replace(
+            {"scan_id": scandc.scan_id},
+            dc.asdict(scandc),
+            upsert=True,
+            return_document=ReturnDocument.AFTER,
+        )
+        if not doc:
+            raise web.HTTPError(
+                500,
+                log_message=f"Failed to insert/update {coll} document ({scandc.scan_id})",
+            )
+        scandc = from_dict(type(scandc), doc)
+        LOGGER.debug(f"replaced: ({coll=}) doc {scandc}")
+        return scandc
+
+
+# -----------------------------------------------------------------------------
+
+
+class ManifestClient(ScanIDCollectionFacade):
+    """Wraps the attribute for the metadata of a scan."""
+
+    async def find_one(self, scan_id: str, incl_del: bool) -> schema.Manifest:
+        """Find one manifest and return it."""
+        return await self._find_one(
+            _MANIFEST_COLL_NAME, scan_id, schema.Manifest, incl_del
+        )
+
+    async def upsert(self, scandc: schema.Manifest) -> schema.Manifest:
+        """Insert/update manifest and return it."""
+        return await self._upsert(_MANIFEST_COLL_NAME, scandc)
+
+    # ------------------------------------------------------------------
+
+    async def get(self, scan_id: str, incl_del: bool) -> schema.Manifest:
+        """Get `schema.Manifest` using `scan_id`."""
+        LOGGER.debug(f"getting manifest for {scan_id=}")
+        manifest = await self.find_one(scan_id, incl_del)
+        return manifest
+
+    async def post(self, event_id: str) -> schema.Manifest:
+        """Create `schema.Manifest` doc."""
+        LOGGER.debug(f"creating manifest for {event_id=}")
+        manifest = schema.Manifest(
+            uuid.uuid4().hex,
+            False,
+            event_id,
+            # TODO: more args here
+        )
+        manifest = await self.upsert(manifest)
+        return manifest
+
+    async def patch(self, scan_id: str, progress: dict[str, Any]) -> schema.Manifest:
+        """Update `progress` at doc matching `scan_id`."""
+        LOGGER.debug(f"patching progress for {scan_id=}")
+        if not progress:
+            msg = f"Attempted progress update with an empty object ({progress})"
+            raise web.HTTPError(
+                422,
+                log_message=msg + f" for {scan_id=}",
+                reason=msg,
+            )
+        manifest = await self.find_one(scan_id, True)
+        manifest.progress = progress
+        manifest = await self.upsert(manifest)
```

Comment:
This will suffer from race conditions with the two independent DB calls.

Reply:
Fixed :heavy_check_mark: